### PR TITLE
Add SWEET vetter (updated)

### LINKED ISSUE #21: Sweet vetter
Goal: Bring the `sweet` vetter into `exovetter`.
Submitted by: @mustaric (See below for details.)

https://github.com/exoplanetvetting/DAVE/blob/master/tessPipeline/sweet.py

It was written by @fergalm . The top function is `runSweetTest`  which basically just runs `computeSweetMetrics` which returns a 3x3 array. I think we want `exovetter` to return that 3x3 array. We don’t need to keep the messages. As a result we also don’t need to input `threshold_sigma`.

The period, epoch and duration inputs can be replaced with the `tce` dictionary we created. And the time and flux should be input as a `lightkurve` object, like we did for `lpp`.

### PR DESCRIPTION
**Pull request by pllim**
This superseded #25 . I also applied comments from that PR.

Fix #21

p.s. You can ignore the coverage "failures".

### CODE REVIEW

File: tests/test_sweet.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    lcf = lc_read(filename)
+    lc = lcf.PDCSAP_FLUX
+
+    sweet_vetter = Sweet()
+    sweet_vetter.run(tce, lc)
+
+    res = sweet_vetter.result
+    assert 'OK' in res['msg']
+    assert_allclose(
+        res['amp'], [[5.13524349e+02, 6.28200510e+02, 8.17452934e-01],
```
Comment (pllim) [inline tests/test_sweet.py:37]: Still not sure about this test. See https://github.com/spacetelescope/exovetter/pull/25#discussion_r496238865

File: exovetter/vetters.py [hunk 2]
```diff
@@ -132,16 +119,122 @@ def run(self, tce, lightcurve):
...
+        self.result = None
+        self.lsf = None
+        self.threshold_sigma = threshold_sigma
+
+    def run(self, tce, lightcurve):
+        self.tce = tce
+        self.lc = lightcurve
+        epoch = tce.get_epoch(getattr(exo_const, lightcurve.time_format))
+
+        # TODO: Do we want results to have unit? If so, what?
```
Comment (pllim) [inline exovetter/vetters.py:167]: Not clear to me if you want the results to be `Quantity` or not. If yes, what are the units? Would be nice if result is naturally `Quantity` if we can make the low level functions take `Quantity` but that is significant refactoring work. See #16.

File: exovetter/vetters.py [hunk 3]
```diff
@@ -132,16 +119,122 @@ def run(self, tce, lightcurve):
...
+        axes[1].plot(phase, best_fit, 'r.')
+        axes[0].set_ylabel('Flux')
+        axes[1].set_xlabel('Phase')
+        axes[0].set_title(f'{self.tce.get("target_name", "Target")} '
+                          f'({self.tce.get("event_name", "Event")})')
+        plt.draw()
+
+        return fig
+
+    def _do_fit(self, time, flux, period_in, epoch, duration_in):
```
Comment (fergalm) [inline exovetter/vetters.py:191]: I think it makes sense to break off _do_fit into a separate, sweet.py module. That better separates the interface from the implementation.

File: exovetter/vetters.py (hunk 2 above)
Comment (fergalm) [reply exovetter/vetters.py:167]: This run method understands what is being passed to it (a lightkurve object), and so is the place to deal with issues of Quantities. 

a) If you choose to implement _do_fit as a method that accepts quantities, run() is responsible for ensuring that the values passed are indeed quantities.

b) If you choose to implement _do_fit as a function that expects numpy arrays, then run() should ensure that the arrays passed are of the correct units.

In our specific case here, the implementation returns the amplitude of the out of transit variation in the same units (e.g ppm) as the input lightcurve. it looks like lightkurve stores a numpy array for flux, and has a separate atribute for Unit. In this case, what I would likely do is pass numpy arrays to _do_fit(), then assign units to the outputs in run(). So something like

```
def run(self):
    ...
   flux = lightcurve.flux_quantity
   ...
   result, lsf = ...
    result[:,0] = u.Quantity(result[:,0], lightcurve.flux_unix)
    result[:,1] = u.Quantity(result[:,0], lightcurve.flux_unix)
```

Does this comment resolve all the questions?

File: exovetter/vetters.py [hunk 4]
```diff
@@ -132,16 +119,122 @@ def run(self, tce, lightcurve):
...
+
+        idx = mark_transit_cadences(time, period_days, epoch, duration_days)
+        flux = flux[~idx]
+
+        out = []
+        for per in [period_days * 0.5, period_days, 2 * period_days]:
+            phase = np.fmod(time - epoch + per, per)
+            phase = phase[~idx]
+            period = np.max(phase)
+            f_obj = WqedLSF(phase, flux, None, period=period)
```
Comment (fergalm) [inline exovetter/vetters.py:218]: As you've implemented WqedLsf fits a sine curve to data with a zero mean. But you haven't subtracted off the mean flux from the star in this function before fitting.  Did I miss it? The data from lightkurve is likely to have a mean value around which the flux oscillates.

File: tests/test_sweet.py (hunk 1 above)
Comment (fergalm) [reply tests/test_sweet.py:37]: Could this issue be related to no subtracting the mean (discussed above?)

**Review by fergalm — COMMENTED** [review]
I think I've addressed the quantities question, let me know if this is not sufficient.

I suspect I've found a bug in _do_fit, where the mean flux from the lightcurve is not removed before fitting a sine wave. That may explain the failing test.

Do you find the code coverage notifications within the diffs helpful? I found it made the code harder to read for the purposes of code review.

File: exovetter/vetters.py (hunk 3 above)
Comment (pllim) [reply exovetter/vetters.py:191]: Sure, in that case, might as well rename `utils.py` to `sweet.py`. Yes, I can do that.

File: exovetter/vetters.py (hunk 4 above)
Comment (pllim) [reply exovetter/vetters.py:218]: I really don't know all the science behind the implementation. What I have was translated from to the best of my ability:

https://github.com/exoplanetvetting/DAVE/blob/test_DAVE/tessPipeline/sweet.py

https://github.com/exoplanetvetting/DAVE/blob/test_DAVE/fileio/kplrfits.py

https://github.com/exoplanetvetting/DAVE/blob/test_DAVE/misc/lsf.py

Given that DAVE has no test for this, I was unable to vet my ported code properly.

Can you please suggest a fix using the "Insert a suggestion" feature? Thanks.

**Comment by pllim** [comment]
> Do you find the code coverage notifications within the diffs helpful?

I actually do. I find it more helpful than the status checks (which I have learned to just ignore).

File: exovetter/vetters.py (hunk 3 above)
Comment (fergalm) [reply exovetter/vetters.py:191]: Make a separate sweet.py. There are some functions in utils.py that will be generally useful (e.g mark_transit_cadences)

File: exovetter/vetters.py (hunk 4 above)
Comment (fergalm) [reply exovetter/vetters.py:218]: I'll have a look at it. I think the DAVE code assumes the mean has already been subtracted/

File: exovetter/vetters.py (hunk 3 above)
Comment (pllim) [reply exovetter/vetters.py:191]: Sure. I'll do that along with the other changes once I hear back from you whether to subtract mean or not. Thanks!

**Comment by fergalm** [comment]
I had a look at this last night. There was a bit more to do than I thought (DAVE does work to the lightcurve before it gives it to the sweet test, we don't). I got close to something I could push, but it'll need another evening of my time.

**Comment by pllim** [comment]
No rush on my end. Thanks for looking into it, @fergalm !

**Pull request closed without merging**